README.md

# congestion18tynewear

<!-- badges: start -->
[![Travis build status](https://travis-ci.org/ppintosilva/congestion18tynewear.svg?branch=master)](https://travis-ci.org/ppintosilva/congestion18tynewear)
<!-- badges: end -->

This package contains data of several events that affected 
traffic flow in the region of Tyne and Wear in the year 2018:

- `metadata` contains information about the events as tweeted by the regional
[Urban Traffic Management & Control facility](https://twitter.com/NELiveTraffic).
- `events` contains the following data for each event:
  - traffic `flow` captured around the epicentre of the traffic event,
  within a time window of several hours before and after the incident.
  - `spatial` features cropped to the area of interest.
  Includes data about monitored locations, shortest paths between
  locations, the enclosing primary (A,B and C roads) and arterial road networks
  and nearby amenities.
  - a `network` capturing the relationships between flows.
  
The main goal of this package is to provide resources to study and profile
traffic congestion under different urban scenarios, using a type of data
which is often not readily available to researchers but whose underlying 
technology is becoming more widespread to monitor traffic flows
within cities - **Automatic Number Plate Recognition**.

The data are ready to be used with the
[anprflows](https://github.com/ppintosilva/anprflows) package
(which was also used to crop it).
Refer to its vignettes for examples on how to do visualise and 
manipulate the data.

## A note on traffic flow data

Flow data refers to origin-destination (OD) traffic flows between pairs of
locations in the road network. This includes the number, and speed statistics 
of vehicles passing between each pair of locations during each time period.
At these locations,
one or more Automatic Number Plate Recognition (ANPR) cameras have been deployed,
so that individual and aggregate travel times can be recorded and inform
traffic bodies of real-time traffic state and network performance.

Flow data is derived from raw ANPR data – a table containing the columns
`vehicle_id | location | timestamp` – using the
[anprx](https://github.com/ppintosilva/anprx) python package.
A link to the associated research paper and underlying methodology is hopefully
coming soon. 
These can then be further spatially cropped, visualised and analysed using the
[anprflows](https://github.com/ppintosilva/anprflows).

## Installation

The package is not yet available in CRAN.

You can install the development version from Github:

``` r
devtools::install_github("ppintosilva/congestion18tynewear")
```


NEWS.md
# congestion18tynewear (development version)

# congestion18tynewear 0.1.1

* Add approximate points in the road network where congestion 
incidents occurred.
* Fill in missing event data documentation.

# congestion18tynewear 0.1.0

* Add data for 7 events of congestion ocurring in 2018 in the region
of Tyne and Wear UK, identified from @NELiveTraffic tweets.
* Underlying traffic flow data in initial development version (no public
release of data on zenodo as of this time yet).

# congestion18tynewear 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(congestion18tynewear)

test_check("congestion18tynewear")


tests/testthat/test-events.R
test_that("events follows an expected structure", {
  expect_true(
    length(metadata) == length(events)
  )

  expected_names <-
    c("nodes", "flows_l", "flows_od", "spatial", "network", "hparams")

  lapply(
    events,
    function(x) names(x) == expected_names
  )
})


tests/testthat/test-metadata.R
test_that("metadata is available", {
  testthat::expect_true(
    length(metadata) > 0
  )
})


R/data.R
#' @importFrom tibble tibble
NULL

#' Data describing several hand-picked events that affected
#' traffic flow in the region of Tyne and Wear in the year 2018:
#'
#' @format A tibble containing the following columns:
#' \describe{
#' \item{\code{id}}{event id}
#' \item{\code{date}}{event date}
#' \item{\code{start_datetime}}{traffic flow temporal window start time}
#' \item{\code{end_datetime}}{traffic flow temporal window end time}
#' \item{\code{event_time}}{when the event was reported on twitter}
#' \item{\code{tweet}}{tweet content}
#' \item{\code{link}}{url link to the original tweet}
#' }
#'
"metadata"

#' Event data for hand-picked events that affected
#' traffic flow in the region of Tyne and Wear in the year 2018.
#'
#' Includes traffic flow data acquired from a network of Automatic
#' Number Plate Recognition (ANPR) cameras and computed via the
#' anprx python package. Using the anprflows R package, a spatio-temporal
#' window is used to capture relevant traffic and spatial data
#' around the epicentre of the traffic event.
#'
#' @format A list of events, each containing the following data:
#' \describe{
#' \item{\code{nodes}}{ids of locations potentially affected by the incident}
#' \item{\code{flows_l}}{tibble containing the number of vehicles passing
#' through each of the locations during each time period}
#' \item{\code{flows_od}}{tibble containing the number, speed statistics
#' and other information of vehicles passing between each pair of locations
#' during each time period}
#' \item{\code{spatial}}{spatial data composed of several sf dataframes,
#' cropped to the area of interest. Includes data about monitored locations,
#' shortest paths between locations, the enclosing arterial
#' and primary road networks and nearby amenities}
#' \item{\code{network}}{asymptotic flow network computed over the chosen
#' spatio-temporal window}
#' \item{\code{hparams}}{values of the hyperparameters used to identify spurious
#' flows and crop spatial features}
#' \item{\code{incident}}{approximate point where incident or congestion was
#' reported by traffic authorities}
#' }
#'
"events"


data-raw/metadata.R
## code to prepare `event-metadata` dataset goes here
library(tidyverse)
library(lubridate)
library(usethis)

tweet_links <- paste0("https://twitter.com/NELiveTraffic/status/",
                      c(
                        "966227277393682433",
                        "992304917493645312",
                        "1049349986427834368",
                        "1058043340170637313",
                        "1066692134139568128",
                        "986633195352416256",
                        "1029381970248298497"
                      ))

dates <- c(
  "2018-02-21",
  "2018-05-04",
  "2018-10-08",
  "2018-11-01",
  "2018-11-25",
  "2018-03-16",
  "2018-08-14"
)

event_time <- c(
  "08:25",
  "08:28",
  "18:25",
  "17:07",
  "13:56",
  "16:49",
  "15:59"
)

start_times <- c(
  "04:00",
  "04:00",
  "12:00",
  "12:00",
  "06:00",
  "12:00",
  "12:00"
)

end_times <- c(
  "12:00",
  "12:00",
  "22:00",
  "22:00",
  "20:00",
  "20:00",
  "20:00"
)

tweets <- c(
  paste(
    "BROKEN DOWN VEHICLE",
    "A184 Felling Bypass westbound after A195 Lingey Lane #Wardley.",
    "A Land Rover is blocking lane 1 awaiting recovery, police on-scene.",
    "Expect heavy congestion.",
    sep = " "),
  paste(
    "A690 Durham Road, traffic starting to break up after earlier",
    "congestion between the North Moor Lane roundabout and the junction",
    "with B1405 Springwell Road  in #Sunderland",
    sep = " "),
  paste(
    "CONGESTION",
    "B1318 Great North Road northbound through #Gosforth from A189",
    "Blue House to Broadway roundabout is 34 mins.",
    sep = " "),
  paste(
    "CONGESTION",
    "A1018 Wearmouth Bridge & North Bridge Street northbound",
    "in #Sunderland this evening.",
    sep = " "),
  paste(
    "B1318 Barras Bridge & B1307 Percy Street southbound #Newcastle city",
    "centre into Eldon Square car park. ",
    sep = " "),
  paste(
    "B1318 Great North Road, heavy slow moving traffic moving southbound",
    "into #Gosforth with congestion at the A1 - A1056 merge #Newcastle, 4:50pm",
    sep = " "),
  paste(
    "Police are attending an overturned vehicle on the A1 WBP southbound ",
    "near junction 66 Angel of The North #Gateshead.",
    "CLEARED",
    "Previous overturned vehicle A1 Southbound J66 A167",
    "#AngeloftheNorth - J65 A194(M) #Washington now removed.",
    "All lanes open although there is still congestion which will take ",
    "time to clear.",
    sep = " ")
)

metadata <-
  tibble(
    date = lubridate::ymd(dates),
    start_datetime = lubridate::ymd_hm(paste0(dates,start_times)),
    end_datetime = lubridate::ymd_hm(paste0(dates,end_times)),
    event_time = lubridate::hm(event_time),
    tweet = tweets,
    link = tweet_links
  ) %>%
  arrange(date) %>%
  mutate(id = 1:nrow(.)) %>%
  select(id, everything())


write_csv(metadata, "data-raw/metadata.csv")

usethis::use_data(metadata, overwrite = TRUE)


data-raw/events.R
library(tidyverse)
library(lubridate)
library(anprflows)
library(tidygraph)
library(usethis)
library(sf)

# To be defined when adding new data ----

# Event case by case filtering functions
location_filter_functions <- list(
  "2018-02-21" = function(x)
    filter(x, !!quo(ref) == "A184" & !!quo(direction) == "W")
  ,
  "2018-03-16" = function(x)
    filter(x, !!quo(ref) == "B1318" & !!quo(direction) == "S")
  ,
  "2018-05-04" = function(x)
    filter(x, !!quo(ref) == "A690" & str_detect(!!quo(address), "Durham"))
  ,
  "2018-08-14" = function(x)
    filter(x, str_detect(!!quo(address), "Angel of the North"))
  ,
  "2018-10-08" = function(x)
    filter(x, !!quo(ref) == "B1318" & !!quo(direction) == "N")
  ,
  "2018-11-01" = function(x)
    filter(x, !!quo(ref) == "A1018" & str_detect(!!quo(address), "Sunderland"))
  ,
  "2018-11-25" = function(x)
    filter(x, (!!quo(ref) == "B1318" | !!quo(ref) == "B1307") &
             !!quo(direction) == "S")
  #
  # @ADD function for new event here
  #
)

# approximate incident location
incident_points <- tibble::tribble(
  ~lon, ~lat,
  -1.537728, 54.946661,
  -1.623651, 55.030819,
  -1.421769, 54.885849,
  -1.584383, 54.911493,
  -1.621560, 55.014966,
  -1.382897, 54.909823,
  -1.612634, 54.979322
  #
  # @ADD approximate incident/congestion point for new event here
  #
) %>%
  dplyr::mutate(
    geometry = purrr::pmap(list(lon,lat),
                           function(x,y) sf::st_point(c(x,y)))
  ) %>%
  tibble::rowid_to_column(var = "id") %>%
  dplyr::select(id, geometry)


# Functions ----

gen_event_data <- function(
  event_id,
  filename,
  start_datetime,
  end_datetime,
  spatial,
  potential_sites,
  incident_point,
  hparams = list(
    "rate_o" = .1,
    "rate_d" = .1,
    "bbox_margin" = c(-500,-500,500,500)
  )
) {
  # week number
  flows <- read_data(filename)

  print(paste0("Processing event ", event_id, "..."))

  asympt_flows_l <-
    flows %>% anprflows::get_flows_l(by_period = FALSE)

  asympt_flows_od <-
    flows %>% anprflows::get_flows_od(asympt_flows_l, by_period = FALSE)

  asympt_network <- anprflows::flow_network(
    asympt_flows_od,
    spurious_if_below =
      c("rate_o" = hparams[["rate_o"]], "rate_d" = hparams[["rate_d"]]))

  asympt_subnetwork <- anprflows::get_neighbors(
    network = asympt_network,
    nodes = potential_sites$id %>% as.character()
  )

  node_subset <-
    asympt_subnetwork %>%
    tidygraph::activate(nodes) %>%
    as_tibble() %>%
    pull(name) %>%
    as.character()

  flows_subset <-
    flows %>%
    filter(t > start_datetime & t < end_datetime) %>%
    filter(
      o %in% node_subset,
      d %in% node_subset
    )

  flows_l <- flows_subset %>% anprflows::get_flows_l()
  flows_od <- flows_subset %>% anprflows::get_flows_od(flows_l)

  subspatial <-
    anprflows::crop_spatial(
      flows_od,
      spatial,
      bbox_margin = hparams[["bbox_margin"]]
    )

  incident_point <-
    incident_point %>%
    sf::st_as_sf()

  st_crs(incident_point) <- 4326

  # project point to utm
  incident_point <- incident_point %>%
    sf::st_transform(crs = sf::st_crs(spatial$locations))

  event <- list(
    "nodes" = node_subset,
    "flows_l" = flows_l,
    "flows_od" = flows_od,
    "spatial" = subspatial,
    "network" = asympt_subnetwork,
    "hparams" = hparams,
    "incident" = incident_point
  )

  return(event)
}

read_data <- function(
  filename
) {
  print(paste0("Reading event data for input file: ", filename))

  flows <- anprflows::read_flows_csv(filename, skip = 1)

  return(flows)
}

get_potential_sites <- function(spatial, filter_function) {
  spatial$locations %>%
    as_tibble %>%
    filter_function %>%
    sf::st_as_sf()
}

# Main loop ----

# local directory where spatial and pipeline data can be found
# FUTURE #TODO -> download from zenodo the necessary files
metadata <- read_csv("data-raw/metadata.csv")

dataset_dir <- "pipeline-data/"
filenames <-
  paste0(
    dataset_dir,
    "flows/fifteenmin/flows_15min_anpr18_week",
    sapply(metadata$start_datetime, lubridate::week),
    ".csv"
  )

spatial <- read_rds(paste0(dataset_dir, "spatial.rds"))

hparams = list(
  "rate_o" = .1,
  "rate_d" = .1,
  "bbox_margin" = c(-500,-500,500,500)
)

potential_sites <-
  lapply(metadata$date %>% as.character(),
         function(d) get_potential_sites(
           spatial,
           location_filter_functions[[d]]))

events <-
  lapply(
    metadata$id,
    function(i)
      gen_event_data(event_id = i,
                     filename = filenames[i],
                     start_datetime = metadata$start_datetime[i],
                     end_datetime = metadata$end_datetime[i],
                     spatial = spatial,
                     potential_sites = potential_sites[[i]],
                     incident_point = filter(incident_points, id == i))
  )

usethis::use_data(events, overwrite = TRUE)
